### PR TITLE
Correct contact bugs sharp

### PR DESCRIPTION
**Pull request by Luckabarbeau**
# Description of the problem

- When particles are in contact using the sharp IB solver, circular reference for the pressure dof.

# Description of the solution

- Check if a cell is cut by more than 1 particle to treat it differently if it's the case

# How Has This Been Tested?

- This as been tested on a case where multiple particles are in contacts (to large to become a test)

# Future changes

- This is the first step in refactoring how the dummy DOFs are handled. More evolved treatment of these special causes should come in the upcoming weeks.

### CODE REVIEW

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 1]
```diff
@@ -1309,7 +1320,7 @@ GLSSharpNavierStokesSolver<dim>::calculate_L2_error_particles()
           bool cell_is_cut;
           // std::ignore is used because we don't care about what particle cut
           // the cell.
```
Comment (oguevremont) [inline source/fem-dem/gls_sharp_navier_stokes.cc:1322]: Should we specify that we don't care how many particles cut the cell?

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 2]
```diff
@@ -1360,7 +1371,7 @@ GLSSharpNavierStokesSolver<dim>::calculate_L2_error_particles()
...
```
Comment (oguevremont) [inline source/fem-dem/gls_sharp_navier_stokes.cc:1373]: Same question here

**Review by oguevremont — COMMENTED** [review]
It looks good to me!

**Review by oguevremont — APPROVED** [review]

File: include/fem-dem/gls_sharp_navier_stokes.h [hunk 3]
```diff
@@ -543,10 +543,11 @@ Return a bool that describes  if a cell contains a specific point
     vertices_to_cell;
   /*
    * This map uses the cell as the key, and store the following information:
-   * if that cell is cut (bool) and what particle cut this cell (unsigned int).
+   * if that cell is cut (bool), what particle cut this cell (unsigned int), and
+   * the number of particles that cut this cell.
    */
   std::map<typename DoFHandler<dim>::active_cell_iterator,
-           std::tuple<bool, unsigned int>>
+           std::tuple<bool, unsigned int, unsigned int>>
```
Comment (blaisb) [inline include/fem-dem/gls_sharp_navier_stokes.h:550]: I don't understand why you now have multiple particles that cut a cell, but you onlystore one particle id. What is that particle id? is it the id of the first prticle? Please be more clear :)

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 4]
```diff
@@ -120,11 +120,12 @@ GLSSharpNavierStokesSolver<dim>::generate_cut_cells_map()
           // this value will never be used.
 
           cell->get_dof_indices(local_dof_indices);
-
+          unsigned int p_count = 0;
```
Comment (blaisb) [inline source/fem-dem/gls_sharp_navier_stokes.cc:123]: What is p_count? Particle count? number_of_particle in cell? Please be more verbose iwth variable name.

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 5]
```diff
@@ -145,30 +146,39 @@ GLSSharpNavierStokesSolver<dim>::generate_cut_cells_map()
...
-                  cell_is_inside = false;
-                  p_id_inside    = 0;
+                  if (p_count == 0)
+                    {
+                      cell_is_cut    = false;
+                      p_id_cut       = 0;
+                      cell_is_inside = false;
+                      p_id_inside    = 0;
+                    }
                 }
```
Comment (blaisb) [inline source/fem-dem/gls_sharp_navier_stokes.cc:178]: Add comments regarding this block of if # else if statements.

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 6]
```diff
@@ -1309,7 +1320,7 @@ GLSSharpNavierStokesSolver<dim>::calculate_L2_error_particles()
           bool cell_is_cut;
           // std::ignore is used because we don't care about what particle cut
```
Comment (blaisb) [inline source/fem-dem/gls_sharp_navier_stokes.cc:1321]: Can you be a bit more verbose about what you mean there.

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 7]
```diff
@@ -2403,7 +2416,9 @@ GLSSharpNavierStokesSolver<dim>::sharp_edge()
           // The id of the particle that cut the cell. Returns 0 if the cell is
```
Comment (blaisb) [inline source/fem-dem/gls_sharp_navier_stokes.cc:2416]: What happens when multiple particles cut a cell?

**Review by blaisb — CHANGES_REQUESTED** [review]
Can you document the changes  a bit more, this could be made clearer

File: source/fem-dem/gls_sharp_navier_stokes.cc (hunk 2 above)
Comment (Luckabarbeau) [reply source/fem-dem/gls_sharp_navier_stokes.cc:1373]: I added the precision

File: source/fem-dem/gls_sharp_navier_stokes.cc [hunk 8]
```diff
@@ -113,18 +113,22 @@ GLSSharpNavierStokesSolver<dim>::generate_cut_cells_map()
         {
           bool         cell_is_cut;
           bool         cell_is_inside;
-          unsigned int p_id_cut    = 0;
-          unsigned int p_id_inside = 0;
-          // is the particle index that cut the cell if it's cut.  If the cell
+          unsigned int p_id_cut                              = 0;
+          unsigned int p_id_inside                           = 0;
```
Comment (blaisb) [inline source/fem-dem/gls_sharp_navier_stokes.cc:117]: use more verbose prefix than p_ please.
e.g. particle_id_which_cuts or something like that. Easier to understand for people reading your code.
<

Comment (Luckabarbeau) [reply source/fem-dem/gls_sharp_navier_stokes.cc]: Perfect I applied this comment

**Review by blaisb — APPROVED** [review]
Great, much clearer now.

**Pull request merged**